Add an `is_chordal` algorithm

Luis-Varona opened the pull request:
We implement Tarjan and Yannakakis (1984)'s MCS algorithm, taking inspiration from the existing NetworkX implementation. Everything is done except for example doctests in `src/chordality.jl` and unit tests in `test/chordality.jl`. (This PR is part of a new suite of algorithms outlined in issue #431.)

Luis-Varona committed:
WIP: Add an is_chordal algorithm

We implement Tarjan and Yannakakis (1984)'s MCS algorithm, taking inspiration from the existing NetworkX implementation. Everything is done except for example doctests in src/chordality.jl and unit tests in test/chordality.jl. (This PR is part of a new suite of algorithms outlined in issue #431.)

simonschoelly commented:
Do you by any change know where I could find a public accessible version of that paper? My usual methods did not help.

simonschoelly commented on src/chordality.jl:
```diff
@@ -0,0 +1,78 @@
+# Examples
+TODO: Add examples
+"""
+function is_chordal(g::AbstractSimpleGraph)
```
```suggestion
function is_chordal(g::AbstractGraph)
```
We can extend this to `AbstractGraph` - we don't support multigraphs anyways currently. `AbstractSimpleGraph` is more of a technical type for historical reasons. As long as you don't modify the graph we probably don't need it.

Furthermore we can use traits to check if a graph is directed or not - for an example see here: https://github.com/JuliaGraphs/Graphs.jl/blob/c001daba3ae32a8649af5e21c0c9786d9c38e082/src/biconnectivity/articulation.jl#L23

simonschoelly replied:
Oh I see you are using `induced_subgraph` below - that might requite a `SimpleGraph` - perhaps there is a way to avoid that? This function is very costly as it needs to allocate a new graph.

Luis-Varona replied:
Thanks for the review, @simonschoelly&mdash;will respond in full later today/tomorrow. I will say for now that (I think?) `AbstractSimpleGraph` rules out the possibility of parallel edges, which would otherwise be computationally expensive to check, but I'm not sure what your thoughts on the matter are.

simonschoelly replied:
I think in order to work with other graph types defined in other packages it is better to use `AbstractGraph` - for example the graphs defined here are not subtypes of `AbstractSimpleGraph`: https://github.com/JuliaGraphs/SimpleWeightedGraphs.jl

simonschoelly replied:
The current `AbstractGraph` interface does not support multiple edges at all - this is something that we have to tackle in the future - but for now the assumption is that all graphs are simple graphs with potential loops.

simonschoelly replied:
So you don't need to worry about that case.

Krastanov commented:
[out.pdf](https://github.com/user-attachments/files/20560769/out.pdf)

The paper is attached here, courtesy of the UMass library.

Luis-Varona commented:
> [out.pdf](https://github.com/user-attachments/files/20560769/out.pdf)
> 
> The paper is attached here, courtesy of the UMass library.

Wonderful, @Krastanov, thanks!

Luis-Varona commented:
@simonschoelly, just wanted to let you know I certainly haven't abandoned this PR. I was busy with work this past week and I'm at a conference this weekend, but I'll get back to it shortly. I've looked over your comments and it shouldn't be many more changes at all anyway, hehe.

Luis-Varona committed:
Fix input type for is_chordal

Originally, we restricted the input type for is_chordal to AbstractSimpleGraph to rule out parallel edges, but some other graph types we would like to support (such as SimpleWeightedGraph) are of the more general AbstractGraph type. It turns out the current AbstractGraph interface aleady does not support parallel edges, so there is no worry here--it is already stated in the Implementation Notes anyway that is_chordal should not take in graphs with parallel edges as input.

Luis-Varona commented:
@simonschoelly, just changed the input type from `AbstractSimpleGraph` to `AbstractGraph`. Anything else you'd like me to change before I get around to adding tests?

Luis-Varona commented:
Wait—I just remembered your comment about induced subgraphs. I'll think about that and get back to you I was largely just porting over networkx's implementation, which does do this, but it probably really isn't optimal.

Luis-Varona committed:
Stop allocating induced subgraphs

Originally mirroring the networkx implementation, we created a new AbstractGraph object every time we tested subsequent neighbors in the potential PEO with the induced_subgraph function. This commit makes our implementation more performant by simply taking the vertex (sub)set and checking to see if all pairs are adjacent via iteration.

Additionally, we change inconsistent naming in certain parts ('node' vs 'vertex'), changing everything to 'vertex' where relevant.

Luis-Varona commented:
@simonschoelly @Krastanov I've fixed the issue regarding unnecessary allocations with `induced_subgraph`, instead having the `_induces_clique` helper take in a vertex subset and, in the context of the original graph, confirm that all possible edges exist. @simonschoelly, is there anything else I should change before I start on unit tests and docstring examples?

Luis-Varona commented:
Hi @simonschoelly, just following up. Is everything good enough for me to conclude my work on the actual logic and start on the docstring and unit tests? 🙂

Luis-Varona commented:
What do you think, @Krastanov? 🙂

Krastanov commented:
Thanks, @Luis-Varona ! No need to wait for anything else to be finished before adding tests on my end. I actually frequently refuse to review PRs before the tests are added, as I judge quality first by the tests and documentation.

I am adding some quality of life improvements to IGraphs.jl that should make it easy to test against their implementation as well. I will post it here shortly.

Krastanov commented:
Just to clarify: while I might have a particular style of review, that style is not shared by everyone, and please do not hesitate to request me approaching this differently.

Krastanov commented:
IGraphs.jl just got a new release in which you can use the following in order to test your implementation against the ones in the igraph C library https://igraph.org/c/html/latest/igraph-Structural.html#igraph_is_chordal

```
the_graph = random_regular_graph(10,5)
LibIGraph.is_chordal(IGraph(the_graph), IGNull(), IGNull(), IGNull(), IGNull())
```

I frequently use random sampling over appropriate ensembles as a way to test that two implementations give the same result. This can be useful here as well.

It can also be fun to compare performance (just make sure to not count the time to convert the graph from one format to the other).

simonschoelly commented:
There is also https://github.com/wangjie212/ChordalGraph

simonschoelly commented:
You resolved that conversation but might have overlooked what I wrote there about using the `IsDirected` trait.

Luis-Varona commented:
Thanks for the info, @Krastanov!

> You resolved that conversation but might have overlooked what I wrote there about using the `IsDirected` trait.

So I did, @simonschoelly. Sorry—I must've overlooked that. I expect that sort of fix shouldn't take long, anyway, so I'll try to take a crack at it later this week.

I suppose regardless of the internal workings of our version of `is_chordal`, I can also start right away on the test suite. Testing against only one of the two libraries suggested above (i.e., `ChordalGraphs.jl` _or_ `IGraphs.jl`, not both) should be sufficient, right?

Krastanov commented:
Yes, no need to test against multiple other libraries. Your own correctness tests are probably most important -- testing against another library with a bunch of random samples just provides a bit more certainty that we are not missing anything.

Luis-Varona commented:
Hi @Krastanov and @simonschoelly, sorry for the long silence&mdash;I plan to return to this issue sometime next week. IIRC, all that's left is just a test suite, and maybe a docs fix. 🙂

Krastanov commented:
Thanks @Luis-Varona

samuelsonric commented:
@Luis-Varona @Krastanov @simonschoelly  I have an extremely performant implementation of this function [here](https://github.com/AlgebraicJulia/CliqueTrees.jl/blob/ccb384de122d73230b39158837d97bcce9cc73ad/src/chordal_graphs.jl#L6), if that is of any interest. In particular, the MCS algorithm is best implemented using a bucket queue: see [here](https://github.com/AlgebraicJulia/CliqueTrees.jl/blob/ccb384de122d73230b39158837d97bcce9cc73ad/src/elimination_algorithms.jl#L1487).

Luis-Varona committed:
Edit 'is_chordal': use trait dispatch, add docstring examples

Luis-Varona committed:
Add chordality docs page

We create docs/src/algorithms/chordality.md, adding it to the Algorithms API section of docs/make.jl.

Luis-Varona committed:
Add IGraphs as test dependency

We add IGraphs as a test dependency (to the nested Project.toml in the test folder, and import it as well in runtests) for our soon-to-be-completed chordality test suite.

Luis-Varona committed:
Fix is_chordal for GenericGraph vertex types

The GenericGraph type used in the unit tests returns a generator object when neighbors is called on it. Originally, we computed the subsequent_neighbors variable via intersect(neighbors(g, v), numbered), but intersect could not infer the element type of a generator object and fell back to Vector{Any}. This caused a MethodError when passing the result to _induces_clique, so we replaced the intersect call with a filter call.

Luis-Varona committed:
Refactor 'is_chordal' (and helpers) for clarity

Short-circuit one-liners (e.g., 'A || do B') are expanded into 'if' blocks (similarly, 'if !A; do B; end'), and the cardinality closure in '_max_cardinality_vertex' is inlined as a lambda function.

Luis-Varona commented:
@Krastanov, I've finished the test suite locally and an just adding some comments for clarity before pushing. I'll probably finish it either before sleeping or sometime tomorrow. :slightly_smiling_face:

(And of course, I'll update the CHANGELOG as well once I'm done with that.)

Luis-Varona committed:
Update is_chordal docstring

Update the example in the is_chordal docstring to construct a chorded 4-cycle more idiomatically.

Luis-Varona committed:
Finish chordality tests

The test suite in src/chordality.jl for the is_chordal method is now complete.

Luis-Varona committed:
Add CHANGELOG entry for is_chordal

Since we have just added the is_chordal method, the CHANGELOG has been updated accordingly to finish up the PR.

Luis-Varona commented:
@Krastanov (and @simonschoelly if you're still interested in this PR), everything should be good and ready for review now! :slightly_smiling_face:

Luis-Varona commented:
@Krastanov (and @simonschoelly if you're still interested in this PR), everything should be good and ready for review now! :slightly_smiling_face:

Krastanov committed:
Merge branch 'master' into 431-is_chordal-algorithm

Krastanov commented:
Thank you for the implementation and apologies for the slow response.

This looks great, I have only a minor question about the docstring. The docstring says "linear time" but my naive reading is the following:

```
while ... # on each iteration we take a vertex from unnumbered and put it in numbered
    _max_cardinality_vertex(...) # complexity of this line is >= length(numbered)*length(unnumbered)
```

so we have `complexity = number of iterations * bound for length of numbered * bound for length of unnumbered`

Isn't this n^3 complexity?

Otherwise this looks good to merge to me (and we can then politely ask @samuelsonric to add dispatch to their algorithm in their package)

Krastanov commented on docs/src/algorithms/chordality.md:
```diff
@@ -0,0 +1,17 @@
+# Degeneracy
```
Should the title be `Chordality` instead of `Degeneracy`?

Luis-Varona replied:
Oops, I'm fairly certain I copy-pasted that from another docs file to get the same structure! I'll fix it once I'm back at my computer.
